### PR TITLE
[BB-1337] Initial implementation

### PR DESCRIPTION
**Pull request by lgp171188**
Testing instructions are in the `README.md` file.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+GitHub Webhook setup
+====================
+
+* Use a tool like `ngrok` to expose the local development server on the internet.
+* Create a webhook on the repository to be watched by following [these instructions](https://developer.github.com/webhooks/creating/).
+This requires administrator-level access to the repository.
+* Use the `<ngrok URL>/pull-requests` as the payload URL and set the `Content Type` to `application/json`.
+* Remember to set a secret and to select only the 'Pull request' event.
+* Create a GitHub web token.
```
Comment (pomegranited) [inline README.md]: nit: Please mention that no particular permissions are required for a public repo, and list the permissions that are required for private repos.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+Pre-requisites
+==============
+
+Python 3.6 or newer
+
+
+GitHub Webhook setup
+====================
+
+* Use a tool like `ngrok` to expose the local development server on the internet.
```
Comment (pomegranited) [inline README.md]: Is the `ngrok` step here just to make the service accessible from the internet?  If so, I think it's sufficient to say, "Run the service on an internet accessible URL.  In development, you can use a tool like `ngrok` to expose your local development server."

File: requirements.txt [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Flask==1.0.3
+PyGithub==1.43.7
```
Comment (pomegranited) [inline requirements.txt:3]: Missing one:
```
flask-mail==0.9.1
```

File: settings.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+GITHUB_WEBHOOK_SECRET = os.environ['GITHUB_WEBHOOK_SECRET']
+GITHUB_ACCESS_TOKEN = os.environ['GITHUB_ACCESS_TOKEN']
+
+WATCH_CONFIG = {
+    'edx/edx-platform': {
+        'patterns': ['doc/*', ],
+        'recipients': ['nobody@example.com', ],
+        'subject': '[ADR] Pull request: {repo} #{number} {action}',
+    },
+}
```
Comment (pomegranited) [inline settings.py]: Since we're going to want to run this service without changing this file, could you make the `WATCH_CONFIG` (or at least its parts) also configurable from the env?

Then, you can also test these overrides without requiring a `test_settings.py`, by [patching os.environ](https://stackoverflow.com/a/31583528/4302112): 
```python
k = mock.patch.dict(os.environ,{'mytemp':'mytemp'})
```

File: pr_watcher_notifier/github_api.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                        break
+                if matched:
+                    break
+        if matched:
+            notify = True
+            if action == 'synchronize':
+                target = get_target_branch(pr)
+                previous_head = data['before']
+                for modified_file in get_comparison_file_names(repo, target, previous_head):
+                    if fnmatch(modified_file, 'docs/*'):
```
Comment (pomegranited) [inline pr_watcher_notifier/github_api.py]: Bug: need use the configured pattern(s) here too.

Your [`test_pr_synchronized_but_already_notified`](https://github.com/open-craft/pr_watcher_notifier/pull/1/files#diff-ee7175cd85f3b81c0c9c02ebba5283c6R131) didn't catch this because your [tests also match the default `docs/*`](https://github.com/open-craft/pr_watcher_notifier/pull/1/files#diff-2b5bde3dd5ae5bc227010549100ba6a8R6), so maybe change that too?

File: pr_watcher_notifier/templates/email_body.txt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+Hi,
+
+Pull request #{{ number }} ({{ pr_url }}) against the {{ repo }} repository has been {{ action }}.
```
Comment (pomegranited) [inline pr_watcher_notifier/templates/email_body.txt:3]: Could you add (all) the [matching files](https://github.com/open-craft/pr_watcher_notifier/pull/1/files#diff-fb78b20e3c0e71c3361c7925c2cded42R79) from the webhook to the message template?

And add some text like, "This PR contains files matching {{ pattern }}."

File: pr_watcher_notifier/github_api.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        for modified_file in pr.get_files():
+            config = current_app.config['WATCH_CONFIG'].get(repo)
+            if config:
+                for pattern in config['patterns']:
+                    if fnmatch(modified_file.filename, pattern):
+                        matched = True
+                        break
+                if matched:
+                    break
+        if matched:
```
Comment (pomegranited) [inline pr_watcher_notifier/github_api.py:86]: Worth adding a comment explaining why what this block of code achieves, something like "Avoid sending multiple notifications out for a given PR by checking the previous commit for files matching `pattern`."

File: README.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+```bash
+$ python3 -m venv venv
+$ source venv/bin/activate
+$ pip install -r requirements.txt -r requirements-dev.txt
+```
+
+* Verify that the unit tests run without errors and pass.
+
+```bash
+$ pytest
```
Comment (pomegranited) [inline README.md]: Could you make this `pytest --cov=pr_watcher_notifier` instead?  The current coverage (90%) is sufficient, but it's useful to see.

File: README.md [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+Python 3.6 or newer
+
+
+GitHub Webhook setup
+====================
+
+* Use a tool like `ngrok` to expose the local development server on the internet.
+* Create a webhook on the repository to be watched by following [these instructions](https://developer.github.com/webhooks/creating/).
+This requires administrator-level access to the repository.
+* Use the `<ngrok URL>/pull-requests` as the payload URL and set the `Content Type` to `application/json`.
```
Comment (pomegranited) [inline README.md]: ...and then this doesn't need to say `ngrok`.

File: README.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+* Configure the environment variables used in the `settings.py` file.
+* Configure the secrets, repositories to watch, the file patterns, the email subject and the recipients in the same file.
+* Set the `FLASK_APP` environment variable to point to the `run.py` file.
+* Set the `FLASK_DEBUG` environment variable to `1` and run the development server.
+
+```bash
+$ export FLASK_APP=run.py
+$ export FLASK_DEBUG=1
+$ flask run
```
Comment (pomegranited) [inline README.md:52]: To use the app on my non-localhost domain, I had to use `flask run --host=0.0.0.0`.  Is there any way to make that IP configurable from the settings or env?

**Review by pomegranited — CHANGES_REQUESTED** [review]
Hi @lgp171188 , looking great for a first implementation!  So glad you were able to figure out a way to avoid sending multiple notifications.

Found one bug, one missing requirement, and left some more comments/questions which you can address at your discretion.  But it's working with my test PR and all the edge cases I threw at it, so will approve when you're ready!

File: settings.py (hunk 4 above)
Comment (lgp171188) [reply settings.py]: @pomegranited, I was planning to do that, but since I was wary of the timebox at that time, I decided against doing it. But since I have time and this shouldn't take a lot of time, I will fix this.

File: pr_watcher_notifier/github_api.py (hunk 5 above)
Comment (lgp171188) [reply pr_watcher_notifier/github_api.py]: @pomegranited, thanks for catching this!

File: pr_watcher_notifier/templates/email_body.txt (hunk 6 above)
Comment (lgp171188) [reply pr_watcher_notifier/templates/email_body.txt:3]: The current code uses only the first match to make a decision because I thought that it will be efficient. Thinking about it now, I don't think the performance difference is going to matter unless we have thousands of files modified. So I'll make the change.

File: README.md (hunk 10 above)
Comment (lgp171188) [reply README.md:52]: @pomegranited, the `flask run` command doesn't read from settings or the environment variables. The old way of writing a file with `app.run(host='0.0.0.0')` in the `if __name__ == '__main__'` block of a script file is not recommended any more.

If there is time, I can create a script `run_dev.sh` reads the environment variable if it is set and modifies the command appropriately. Otherwise, we might have to stop with documenting this.

**Comment by lgp171188** [comment]
@pomegranited, I have updated the PR to address the review comments. Since I have added `PyYAML` to the requirements file, you will have to install it before testing.

File: README.md (hunk 10 above)
Comment (pomegranited) [reply README.md:52]: 👍

File: settings.py (hunk 4 above)
Comment (pomegranited) [reply settings.py]: Thank you!  Works great.

**Review by pomegranited — APPROVED** [review]
Works great, thanks @lgp171188 ! :+1:

- [x] I tested this by running the service on my dev server according to the README instructions, configured to respond to a webhook for PR against a github fork.
     * Opened PR with no files matching pattern (no email).
     * Added file which did not match pattern (no email).
     * Moved above file so that it did match pattern (email sent).
     * Played with file matching and force push changes to PR branch (emails sent/not sent as expected).
- [x] I read through the code
- [x] I checked for accessibility issues -- email is text, so legible by screenreaders.
- [x] Includes documentation.

File: pr_watcher_notifier/github_api.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+
+def get_target_branch(pr):
+    """
+    Return the target branch name of a pull request.
+    """
+    return pr.base.ref
+
+
+def should_send_notification(data):
```
Comment (nedbat) [inline pr_watcher_notifier/github_api.py:63]: A nit, but this function is not part of "the github api", it's got significant business logic in it.

File: pr_watcher_notifier/github_api.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+            # in the update. This can be done by comparing the previous HEAD of the PR branch against
+            # the target branch and verifying that no files matching the patterns were modified.
+            if action == 'synchronize':
+                target = get_target_branch(pr)
+                previous_head = data['before']
+                for modified_file in get_comparison_file_names(repo, target, previous_head):
+                    for pattern in watch_config[repo]['patterns']:
+                        if fnmatch(modified_file, pattern):
+                            notify = False
+                            break
```
Comment (nedbat) [inline pr_watcher_notifier/github_api.py:100]: Is this the right logic?  It amounts to, "Don't notify if any interesting file in this commit was modified in a previous commit for this pull request."   Shouldn't it be, "Don't notify if there's no interesting file in this commit that wasn't modified in a previous commit"?

Comment (lgp171188) [reply pr_watcher_notifier/github_api.py:100]: @nedbat, the idea was to notify for the first interesting change in a PR and let people follow the PR for future updates. However, this doesn't work well when multiple unrelated watch patterns are added for a repository and a notification is sent only for the first match and not for the subsequent matches for the other patterns which might be interesting for a different set of users.

`Don't notify if there's no interesting file in this commit that wasn't modified in a previous commit` approach will work for the scenario mentioned above at the cost of more email notifications, which is something we wanted to keep to a minimum.

The app treats all the watch patterns for a repository as a single entity (the PR either matches the patterns or not, we don't care about which patterns were matched) and that is why it is not possible to do anything different than what we had in mind at the time of writing this.

However, the concerns you have raised are valid and we are open to addressing them.

CC @pomegranited

File: pr_watcher_notifier/github_api.py (hunk 11 above)
Comment (lgp171188) [reply pr_watcher_notifier/github_api.py:63]: @nedbat, you're right! I will move this to a more appropriate place in an update.
